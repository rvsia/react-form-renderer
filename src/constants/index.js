export const components = {
  TEXT_FIELD: 'text-field',
  TEXTAREA_FIELD: 'textarea-field',
  FIELD_ARRAY: 'field-array',
  SELECT_COMPONENT: 'select-field',
  FIXED_LIST: 'fixed-list',
  CHECKBOX: 'checkbox',
  SUB_FORM: 'sub-form',
  RADIO: 'radio',
  TABS: 'tabs',
  TAB_ITEM: 'tab-item',
  DATE_PICKER: 'date-picker',
  TIME_PICKER: 'time-picker',
  TAG_CONTROL: 'tag-control',
  WIZARD: 'wizard',
};
  /** Validator functions placeholder */
export const validators = {
  REQUIRED: 'required-validator',
  /**
     * min length of the input value
     */
  MIN_LENGTH: 'min-length-validator',
  /**
     * max length of the input value
     */
  MAX_LENGTH: 'max-length-validator',
  /**
   * exact length of input value
   */
  EXACT_LENGTH: 'exact-length-validator',
  /**
     * minimum count of fileds in some dynamic list of fields
     */
  MIN_ITEMS_VALIDATOR: 'min-items-validator',
  /**
     * Minimum value of number input
     */
  MIN_NUMBER_VALUE: 'min-number-value',
  /**
     * Maximum value of number inpuy
     */
  MAX_NUMBER_VALUE: 'max-number-value',
  /**
     * Regexp pattern validator
     */
  PATTERN_VALIDATOR: 'pattern-validator',
};

export const layoutComponents = {
  FORM_WRAPPER: 'FormWrapper',
  BUTTON: 'Button',
  COL: 'Col',
  FORM_GROUP: 'FormGroup',
  BUTTON_GROUP: 'ButtonGroup',
  ICON: 'Icon',
  ARRAY_FIELD_WRAPPER: 'ArrayFieldWrapper',
  HELP_BLOCK: 'HelpBlock',
};
