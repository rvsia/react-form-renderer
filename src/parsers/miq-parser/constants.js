import { components } from '../../constants';

export const componentMap = {
  DialogFieldTextBox: components.TEXT_FIELD,
  DialogFieldRadioButton: components.RADIO,
  DialogFieldCheckBox: components.CHECKBOX,
  DialogFieldTextAreaBox: components.TEXTAREA_FIELD,
  DialogFieldDropDownList: components.SELECT_COMPONENT,
  DialogFieldDateControl: components.DATE_PICKER,
  DialogFieldDateTimeControl: components.TIME_PICKER,
  DialogFieldTagControl: components.TAG_CONTROL,
};

// for true values
export const neededAttributes = [
  'name',
  'label',
  [ 'label', 'title' ],
  [ 'data_type', 'dataType' ],
  [ 'required', 'isRequired' ],
  [ 'read_only', 'isReadOnly' ],
  [ 'description', 'helperText' ],
];

// for false values
export const neededFalseAttributes = [
  [ 'visible', 'isVisible' ],
];
