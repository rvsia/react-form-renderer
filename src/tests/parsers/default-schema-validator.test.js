import output from '../../demo-schemas/miq-schemas/output';
import React from 'react';
import defaultSchemaValidator from '../../parsers/default-schema-validator';
import { components } from '../../constants/';

describe('Default schema validator', () => {
  let formFieldsMapper;
  beforeEach(() => {
    formFieldsMapper = {
      foo: props => <div>Component</div>,
      invalidComponent: 'baz',
      [components.TABS]: ({ children }) => <div>{ children }</div>,
    };
  });
  it('should fail if input is not a object', () => {
    expect(() => defaultSchemaValidator([])).toThrow();
  });

  it('should fail if input object does not have fields key', () => {
    expect(() => defaultSchemaValidator({})).toThrow();
  });

  it('should fail if input object does fields key that is not array', () => {
    expect(() => defaultSchemaValidator({ fields: {}})).toThrow();
  });

  it('should fail if field item does not have component property', () => {
    expect(() => defaultSchemaValidator({ fields: [{}]})).toThrow();
  });

  it('should fail if field item does not have name property', () => {
    expect(() => defaultSchemaValidator({ fields: [{
      component: 'foo',
    }]}, formFieldsMapper)).toThrow();
  });

  it('should fail if field component property is not in form fields mapper.', () => {
    expect(() => defaultSchemaValidator({ fields: [{
      component: 'blarghs',
      name: 'foo',
    }]}, formFieldsMapper)).toThrow();
  });

  it('should fail if field component from form fields mapper is not a valid React component.', () => {
    expect(() => defaultSchemaValidator({ fields: [{
      component: 'invalidComponent',
      name: 'foo',
    }]}, formFieldsMapper)).toThrow();
  });

  it('should fail if field condition is not correct type.', () => {
    expect(() => defaultSchemaValidator({ fields: [{
      component: 'foo',
      name: 'foo',
      condition: '',
    }]}, formFieldsMapper)).toThrow();
  });

  it('should fail if field condition is missing when key.', () => {
    expect(() => defaultSchemaValidator({ fields: [{
      component: 'foo',
      name: 'foo',
      condition: {},
    }]}, formFieldsMapper)).toThrow();
  });

  it('should fail if field condition is missing is key.', () => {
    expect(() => defaultSchemaValidator({ fields: [{
      component: 'foo',
      name: 'foo',
      condition: { when: 'Foo' },
    }]}, formFieldsMapper)).toThrow();
  });

  it('should fail if field condition when property is not correct type.', () => {
    expect(() => defaultSchemaValidator({ fields: [{
      component: 'foo',
      name: 'foo',
      condition: { when: 123, is: 456 },
    }]}, formFieldsMapper)).toThrow();
  });

  it('should fail if field validate is not an array.', () => {
    expect(() => defaultSchemaValidator({ fields: [{
      component: 'foo',
      name: 'foo',
      validate: {},
    }]}, formFieldsMapper)).toThrow();
  });

  it('should fail if field validate item is not an object.', () => {
    expect(() => defaultSchemaValidator({ fields: [{
      component: 'foo',
      name: 'foo',
      validate: [ '' ],
    }]}, formFieldsMapper)).toThrow();
  });

  it('should fail if field validate item does not have type property.', () => {
    expect(() => defaultSchemaValidator({ fields: [{
      component: 'foo',
      name: 'foo',
      validate: [{}],
    }]}, formFieldsMapper)).toThrow();
  });

  it('should fail if field validate item validator type does not exist.', () => {
    expect(() => defaultSchemaValidator({ fields: [{
      component: 'foo',
      name: 'foo',
      validate: [{
        type: 'magic',
      }],
    }]}, formFieldsMapper)).toThrow();
  });

  it('should pass validation', () => {
    expect(() => defaultSchemaValidator(output, {
      ...formFieldsMapper, 'sub-form': () => <div />,
      'text-field': () => <div />,
      'textarea-field': () => <div />,
      checkbox: () => <div />,
      radio: () => <div />,
      'select-field': () => <div />,
      'date-picker': () => <div />,
      'time-picker': () => <div />,
      'tag-control': () => <div />,
    })).not.toThrow();
  });
});
