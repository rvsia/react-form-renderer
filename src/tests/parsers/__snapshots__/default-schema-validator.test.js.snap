// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Default schema validator should fail if field component from form fields mapper is not a valid React component. 1`] = `"FormComponent \\"invalidComponent\\" from formFieldsMapper is not a valid React component!"`;

exports[`Default schema validator should fail if field component property is not in form fields mapper. 1`] = `
"
          Component of type \\"blarghs\\" is not present in formFieldsMapper.
          Please make sure \\"blarghs is included in your formFieldsMapper.\\"
          FormFieldsMapper has these values: [foo,invalidComponent,tabs]
        "
`;

exports[`Default schema validator should fail if field condition is missing is key. 1`] = `
"
      Error occured in field definition with name: \\"foo\\".
      Field condition must have \\"is\\" property! Properties received: [when].
    "
`;

exports[`Default schema validator should fail if field condition is missing when key. 1`] = `
"
      Error occured in field definition with \\"name\\" property: \\"foo\\".
      Field condition must have \\"when\\" property! Properties received: [].
    "
`;

exports[`Default schema validator should fail if field condition is not correct type. 1`] = `
"
      Error occured in field definition with name: \\"foo\\".
      Field condition must be an object, received string!
    "
`;

exports[`Default schema validator should fail if field condition when property is not correct type. 1`] = `
"
      Error occured in field definition with name: \\"foo\\".
      Field condition property \\"when\\" must be oof type \\"string\\", number received!].
    "
`;

exports[`Default schema validator should fail if field item does not have component property 1`] = `"Each fields item must have \\"component\\" property!"`;

exports[`Default schema validator should fail if field item does not have name property 1`] = `"Each fields item must have \\"name\\" or \\"key\\" property! Name is used as a unique identifier of form fields."`;

exports[`Default schema validator should fail if field validate is not an array. 1`] = `
"
      Error occured in field definition with name: \\"foo\\".
      Field validate property must be an Array, object received!
    "
`;

exports[`Default schema validator should fail if field validate item is an object and does not have type property. 1`] = `
"
        Error occured in field definition with name: \\"foo\\".
        Field validator at index: 0 does not have \\"type\\" property! Properties received: [].
      "
`;

exports[`Default schema validator should fail if field validate item is an object and validator type does not exist. 1`] = `
"
        Error occured in field definition with name: \\"foo\\".
        Field validator at index: 0 does not have correct \\"type\\" property!
        Received \\"magic\\", expected one of: [required-validator,min-length-validator,max-length-validator,exact-length-validator,min-items-validator,min-number-value,max-number-value,pattern-validator].
      "
`;

exports[`Default schema validator should fail if field validate item is not an object or a function. 1`] = `
"
        Error occured in field definition with name: \\"foo\\".
        Field validator at index: 0 must be an object or a function, string received!
      "
`;

exports[`Default schema validator should fail if input is not a object 1`] = `"Form Schema must be an object, received array!"`;

exports[`Default schema validator should fail if input object does fields key that is not array 1`] = `"Component of type schema must contain \\"fields\\" property of type array, received type: object!"`;

exports[`Default schema validator should fail if input object does not have fields key 1`] = `"Component of type schema must contain \\"fields\\" property of type array, received undefined!"`;
